Added files necessary for generic templating of Goodies

srvsh opened the pull request:
These are backward compatible changes, as in these can be merged before the changes to the DuckPAN App are merged.

(part of work related to [Issue 268 in p5-app-duckpan](https://github.com/duckduckgo/p5-app-duckpan/issues/268))

moollaza commented:
@srvsh I actually have a PR that updates the existing templates. Let's combine the content of those files #1771 with yours. 

/cc @zachthompson

zachthompson commented on template/share/goodie/example/example.css:
```diff
@@ -0,0 +1,4 @@
+.zci--<: $lia_name :> {
+
```
@moollaza anything appropriate you can think to have here as an example?

srvsh commented:
@moollaza Thanks, I rebased my changes on top of master and kept your `example.js` file.

I noticed that you've used `$lia_id` in the template files, but I couldn't find that variable defined in [`New.pm`](https://github.com/duckduckgo/p5-app-duckpan/blob/5c7e0b5f22afb064ef9a11affa2741146dae49b6/lib/App/DuckPAN/Cmd/New.pm#L63-L69). Will that be added sometime in the future? For now, I have replaced occurances of `$lia_name` with `$lia_id` in my changes as well.

zachthompson commented on template/templates.yml:
```diff
@@ -0,0 +1,38 @@
+---
+# Templates for goodies
+#
+# Format:
```
So I'm wondering if we should slightly change the format of this file to make templates reusable:
```
templates:
  pm:
    input: lib/DDG/Goodie/Example.pm
    output: lib/DDG/Goodie/<:$ia_path:>.pm
  minimal pm:
    input: lib/DDG/Goodie/Minimal.pm
    output: lib/DDG/Goodie/<:$ia_path:>.pm
  js:
    input: share/goodie/example/example.js:
   output: share/goodie/<:$lia_id:>/<:$lia_id:>.js
  test:
    input: t/Example.t:
    output: t/<:$ia_path:>.t
template_configs:
  default:
    description: Standard Goodie instant answer
    required:
      - pm
      - test
    optional:
      - js
      - css    
   with javascript:
     description: Goodie with custom javascript
     required:
       - pm
       - js
       - test
     optional:
       - css
  javascript only:
    description: Goodie that does all of its processing in javascript
    required:
      - minimal pm
      - test
      - js
    optional:
      -css
  all:
    required:
      - pm
      - test
        - js
        - css
```
As we discover common configs people use we can easily add them.  We might consider doing a comma separated string (to allow for whitespace) for the `required/optional` sections just to make it a big more succinct:
```
required: minimal pm, test, js
optional: css
```
What do you think @srvsh?  Would this be useful @moollaza?

srvsh replied:
I think this is a good idea @zachthompson and makes it easier to maintain more template configs. Also the inclusion of `description` fields will enable an easy to use menu like interface for users, which is great. In the same vein, we could add a `description` field to the various templates (i.e. `js`, `pm` etc) so that users get a nice message when selecting optional files (eg: _Javascript_ instead of _js_).

> We might consider doing a comma separated string (to allow for whitespace) for the required/optional sections just to make it a big more succinct:

Something like `required: [ minimal pm, test, js ]` would just work in YAML and return an array of strings, so this is surely feasible.

zachthompson replied:
Sounds good @srvsh.  I agree that anywhere we can make it easier to understand what a component is for would be good.  

A simple menu with numbered options would be cool as well (including something like custom.)

srvsh replied:
Thanks @zachthompson. I have made code changes in [this PR](https://github.com/duckduckgo/p5-app-duckpan/pull/279) to support the new templates.yml format.

srvsh commented:
@zachthompson I have updated the `templates.yml` file with the format you suggested, with some minor modifications. Do let me know what you think!

zachthompson commented on template/templates.yml:
```diff
@@ -0,0 +1,49 @@
+---
+# Template definitions for goodies
+
+# Each template represents one input and one output file
```
Let's make a note about `needs_restart`, something like:

The creation of certain templates will need to restart a *running* duckpan in order to be recognized:
* Perl modules
* cheat sheet json, since they are pre-loaded

srvsh replied:
Done, thanks.

zachthompson commented on template/templates.yml:
```diff
@@ -0,0 +1,56 @@
+
+    cheatsheet:
+        description: Cheat Sheet Instant Answer
+        required: [ cheatsheet json ]
```
commas required?

srvsh replied:
Commas are not required in this case since `cheatsheet json`  is the name of one template.

srvsh commented:
@zachthompson I have renamed `cheatsheet json` to `cheatsheet_json` to make it unambiguous. Do you think the `all` template-set is required?

zachthompson commented:
@srvsh yes, `all` would be for really custom instant answer.  Are you having issues with it?

srvsh commented:
@zachthompson Thanks for the confirmation, I'll keep it. No, there's no issue with it. Since `all` is equivalent to selecting `default` and choosing to generate all the optional files, I thought we could drop it.

moollaza commented:
@srvsh this is looking pretty good. The only thing I believe we need to add is an optional Handlebars file. Goodies can use JS, CSS, and/or Handlebars

moollaza commented:
If you'd like I can merge this and you can submit a new PR? Happy either way

srvsh commented:
@moollaza Thanks! I can add the handlebars file to this PR. Will it be the same as the [one for Spice IAs](https://github.com/duckduckgo/zeroclickinfo-spice/blob/master/template/share/spice/example/example.handlebars)?

moollaza commented:
@srvsh Spice seems fine, both CSS and Handlebars are optional.

srvsh commented:
@moollaza Done, I have added the handlebars file from the Spice IA repository here as well. This is how an invocation of the `duckpan new` command looks now in the Goodie repository:
```
$ duckpan new
Creating a new Standard Goodie Instant Answer...
Please enter a name for your Instant Answer:  Bus Routes
Would you like to configure optional templates? [y/N]: y

  1> Javascript
  2> CSS
  3> Handlebars
  4> Javascript, CSS
  5> Javascript, Handlebars
  6> CSS, Handlebars
  7> Javascript, CSS, Handlebars

Choose configuration [1]: 7
Created files:
    lib/DDG/Goodie/BusRoutes.pm
    t/BusRoutes.t
    share/goodie/bus_routes/bus_routes.js
    share/goodie/bus_routes/bus_routes.css
    share/goodie/bus_routes/bus_routes.handlebars
Successfully created Goodie: BusRoutes
```

srvsh commented:
In the last commit, I have added a `subdir_support: false` field for Cheat Sheets since they cannot be created in user-specified sub-directories.

moollaza commented:
@srvsh Thanks! This LGTM :+1:
